### ISSUE TITLE
RDNA2 (gfx1030) field report: AMD backend works; greedy decode not token-stable across mixed CPU/GPU expert tiers

### ISSUE THREAD

**Issue by ab2m** [JustVugg/colibri#510]
## Summary

Field report: the v1.1.0 AMD backend **works on RDNA2** (RX 6900 XT, gfx1030) — we believe this is the first validation outside the RDNA4 card it was developed on. It runs and the VRAM expert tier behaves as designed, but **greedy decode is not token-stable when experts are split across the CPU and GPU tiers**, which differs from the token-exact result reported for RDNA4 in #339/#499.

## Environment

- GPU: AMD Radeon RX 6900 XT 16 GB (RDNA2, gfx1030 — no WMMA), also driving the desktop
- ROCm 7.1.1 / hipcc 7.1.1 (Ubuntu distro packages, `ROCM_HOME=/usr`), kernel 7.0.0-27-generic
- CPU: Ryzen 9 5900X, 60 GB RAM, model on a Corsair MP600 PRO XT NVMe
- Build: dev branch (v1.1.0), `make HIP=1 HIP_ARCH=gfx1030 ROCM_HOME=/usr` — compiles clean (85 host warnings), engine detects the card as `sm_103`
- Model: `mateogrgic/GLM-5.2-colibri-int4-with-int8-mtp` (int4 + int8 MTP heads)
- Requires the `COLI_TEMP` patch from #509 (comment) for any `--temp` run

## What works

- `COLI_CUDA=1 COLI_GPU=0 CUDA_EXPERT_GB=12`, "routed experts only" mode: hot tier pins learned experts (166 → 1,335 tensors across runs as `.coli_usage` grows), thousands of calls served from VRAM.
- Numbers on this disk-bound box (24-token runs, `--topp 0.85`): cold CPU 0.10 tok/s (hit 46%) → GPU 0.38–0.39 tok/s (hit 63–64%, pin up to 36%). Warm CPU alone also reaches ~0.37 tok/s, so on this machine the GPU tier's main effect is the higher hit rate; disk streaming remains the bottleneck, as expected.

## The finding: greedy decode is placement-dependent on the mixed tier

Same binary, same flags (`--temp 0`, `DRAFT=0`, `--topp 0.85`), same prompt, 3-way matrix:

| Comparison | Result |
|---|---|
| CPU-only run 1 vs run 2 | **token-identical** (deterministic) |
| GPU (mixed tier) vs CPU | 1 token differs (~pos 7: "grande" vs "artificial"), then reconverges |
| GPU run 1 vs GPU run 2 | diverge (~pos 9) — the pinned set had grown between runs (1,083 → 1,335 tensors) |

Interpretation: an expert served from VRAM goes through the HIP kernel, the same expert served from RAM goes through the AVX2 kernel; accumulation order differs, and near-tie argmax steps flip. So "placement only decides speed" holds semantically but not bit-wise on this path. Since #339 reports token-exact CPU-vs-GPU on RDNA4 (with experts in VRAM), the question is whether the non-WMMA fallback path gfx1030 takes is expected to be token-exact, or whether the RDNA4 parity relies on WMMA-specific dispatch.

Happy to run the `benchmark_cuda_fixture.py` A/B or any diagnostics you'd like on this card.

**Comment by JustVugg** [comment]
Thank you — as far as we know this is the **first RDNA2 validation** of the AMD backend, and it's a model field report: the 3-way matrix is exactly the right experiment, and your interpretation is correct.

**What's guaranteed and what isn't.** Token determinism holds *within a fixed kernel family and a fixed placement*. The AVX2 and HIP kernels reduce in different orders, so the same expert served from VRAM vs RAM produces last-ULP-different logits; a greedy step whose top-2 margin sits inside that noise flips, and the sequence diverges from there (your "grande" vs "artificial" at pos ~7). The RDNA4 token-exact result in #339 does **not** rely on WMMA — it means that on that fixture no step landed near enough a tie to flip. It's margins, not architecture: mixed-tier bitwise identity was never a structural guarantee on any GPU, RDNA4 included. Your phrasing — *"placement only decides speed" holds semantically but not bit-wise* — is exactly the intended contract; we'll state it explicitly in the docs.

**Your GPU run1-vs-run2 divergence has a specific, testable cause**: the pinned set grew between runs (1,083 → 1,335 tensors), so some experts changed kernel family *between* runs. Prediction: freeze the placement and GPU greedy decode becomes deterministic run-to-run. The learning cache appends to `<snap>/.coli_usage` after each turn, so:

```bash
cp <snap>/.coli_usage /tmp/usage.frozen
# before EACH run:
cp /tmp/usage.frozen <snap>/.coli_usage
```

If two GPU runs with the frozen file still diverge, that's a real bug and we'd very much want that log.

**Where this class of flip actually matters, the engine already handles it**: speculative verify pins draft and verify to one kernel family (`SPEC_PIN`, #163) precisely because near-tie flips break acceptance. On the plain decode path the flip is cosmetic — both tokens are legitimate argmaxes within FP noise.

If you're up for one more run, the frozen-usage A/B is the single data point that separates "expected FP noise" from "real nondeterminism". And with your permission we'll add **RX 6900 XT (gfx1030, no WMMA)** to the validated-hardware list — routed-experts tier, first RDNA2 confirmation.

(The `COLI_TEMP` dependency on the #509 patch and the `sm_103` detection label are noted and tracked separately.)

**Comment by InspiringCode** [comment]
Will this also help with AMD iGPUs like the one from the AI MAX 395 machines? I am waiting so hard to see the CPU bottleneck removed on these affordable power horses to fully utilize their power! ❤️

**Comment by JustVugg** [comment]
Vulkan landed on `dev` — #418 (@steve-m): expert tier + dense + MLA attention on any Vulkan 1.2 driver (Mesa/RADV included), opt-in via `make VK=1` + `COLI_VULKAN=1`. On gfx1030 this is worth comparing against the HIP path you reported — especially since #681 fixed HIP for matrix-core-less cards, so both backends should now build on RDNA2.

**Comment by huppiflupp** [comment]
Counter-datapoint from a third point on the grid, plus one control I think your
run was missing.

Your open question is whether the RDNA4 token-exactness relies on WMMA-specific
dispatch, since gfx1030 has no WMMA. I have **gfx1151 (RDNA3.5, WMMA present)**
— but reached over the **Vulkan** backend, not HIP, since HIP is broken on this
part. Greedy decode stayed token-exact across mixed CPU/GPU expert placement.

### Setup

- AMD Ryzen AI MAX+ 395 (Strix Halo), Radeon 8060S / gfx1151, 128 GB **unified**
- colibri f028d26, GLM-5.2 int4-gs64, `--temp 0`, `DRAFT=0`, fixed prompt
- Three arms, each run twice, 24 tokens; then A and C re-run at 64 tokens

### The control your matrix lacked

Your GPU run 1 vs run 2 diverged *and* the pin set grew 1,083 → 1,335 tensors
between them. Those two runs weren't comparable, so that row can't separate
"placement is non-deterministic" from "the placement changed". I restored
`.coli_usage` from a frozen copy **before every single run**, so all six started
from a byte-identical pin set.

### Result

| comparison | 24 tok | 64 tok |
|---|---|---|
| CPU vs CPU (determinism baseline) | identical | identical |
| Vulkan dense+attn vs itself | identical | — |
| mixed expert tier vs itself | identical | identical |
| CPU vs Vulkan dense+attn | identical | — |
| **CPU-experts vs mixed tier (your question)** | **identical** | **identical** |
| CPU vs full Vulkan tier | identical | identical |

Byte-identical generated text throughout. The arms demonstrably differed:

```
A: (no [VK] lines at all — pure CPU)
C: [VK] expert tier: 3496 hot experts resident (65.98 GB VRAM, top-3496 of history)
```

and they ran at different speeds (0.74 vs 0.90 tok/s over 64 tokens), so this is
not a case of the GPU path silently not engaging.

### What I think this does and does not show

It does **not** refute your measurement — there are two differences at once, and
I can't isolate them on this box:

1. **Different backend.** Vulkan compute shaders, not HIP kernels. The
   accumulation order that flips your argmax lives in the kernel, so a different
   kernel is a different experiment.
2. **One prompt.** Ten runs, but all on the same prompt, so this is a null
   result over one routing path — weaker evidence than the same count spread
   over several domains would be. Yours diverged on one prompt too, which is
   why I would not read either result as settled.
3. **Unified memory.** On this APU "VRAM" and RAM are the same physical memory.
   That changes what is resident, not how the math accumulates — but I'd rather
   name it than have it found later.

What it does narrow is the hypothesis space: token-exactness across mixed
placement is **not** exclusive to RDNA4, and **not** inherently broken on AMD or
on split tiers. So "does parity rely on WMMA-specific dispatch?" now has a third
answer to account for — a WMMA-capable part that is exact over a *non*-WMMA
backend. That points at the specific gfx1030 HIP fallback kernel rather than at
WMMA presence as such.

If it helps, the cheap discriminator on your box: re-run your 3-way matrix with
`.coli_usage` restored before each run. If GPU run 1 vs run 2 then becomes
identical while GPU vs CPU still differs, the finding is a clean per-kernel
accumulation difference; if run 1 vs run 2 still differs with a frozen pin set,
there is a second source of nondeterminism worth chasing separately.

**Comment by crichalchemist** [comment]
Third point on the grid, chosen to separate the two things the first two confound.

@huppiflupp's reading was that a WMMA-capable part being exact over a non-WMMA backend
points at the gfx1030 HIP fallback kernel rather than at WMMA presence as such. The cell
that tests that is **no matrix cores + Vulkan**. That is a Polaris card:

| datapoint | silicon | matrix cores | backend | mixed-tier greedy |
|---|---|---|---|---|
| this issue | gfx1030 | none | HIP | diverged |
| @huppiflupp | gfx1151 | WMMA | Vulkan | exact |
| this run | Radeon Pro 580 (Polaris/GCN4) | none | Vulkan via MoltenVK | **exact** |

### Setup
- upstream `dev` @ `5a0b725`, `colibri.c` (same engine as your report), Vulkan tier from #418
- 2017 iMac, i7-7700K, macOS 13.7.8; Radeon Pro 580 8 GB, **discrete**, MoltenVK 1.4.1
  (Vulkan 1.4.334, loader 1.4.350.1). Not RADV: MoltenVK compiles the SPIR-V to Metal.
- `make colibri VK=1 ARCH=native` + MacPorts libomp/vulkan flags; `-DCOLI_VULKAN` and
  `-march=native` on the compile line, `otool -L` shows libomp + libvulkan.
- The engine warns on every run that there is no host-visible+device-local memory type
  (no Resizable BAR), so tier weights live in host RAM. The shader still runs on the
  Radeon: `[VK] ready: AMD Radeon Pro 580`.
- Fixture: `tools/make_glm_oracle.py --fmt4` (routed experts int4-g64), two prompts —
  the stock one (20 new tokens) and a second prompt run to 29 tokens. Tiny model, so
  read this as a null result over two routing paths, same as the gfx1151 run.

### Your control, kept
The Vulkan tier fills from `.coli_usage` (top-N of history) and the one-shot oracle path
never writes it, so I wrote one history file once and restored it before **every** run
(md5 `0286025f…`): 16 rows, strictly distinct counts, total 1960 so AUTOPIN stays off.
`COLI_TEMP=0 DRAFT=0 USAGE_SAVE=0`. Tier size via `COLI_VK_EXPERTS`: 64 (all 16 experts
resident), 8, 4.

### Result
Every arm, both prompts, byte-identical token ids — one md5 per prompt across CPU x2,
full tier, half tier x2, quarter tier x2:

| prompt | arms | md5 of the token line | match |
|---|---|---|---|
| p1 | CPU, CPU, VK16, VK8, VK8, VK4, VK4 | `519fd01d625ef478717870b9daad9ce8` | 20/20 all |
| p2 | CPU, CPU, VK16, VK8, VK8, VK4, VK4 | `683e62a16e5dd2d149d01207bbd391f0` | 29/29 all |

That the mixed arms really were mixed, from `ROUTE_TRACE` on the CPU arm (routing is
identical in every arm because the tokens are) against the resident set the fill's
descending sort produces:

| prompt | tier | expert calls tier / cpu | routings with one expert in VRAM and the other on CPU |
|---|---|---|---|
| p1 | 8 of 16 | 39 / 85 | **33 of 62** |
| p1 | 4 of 16 | 17 / 107 | 17 of 62 |
| p2 | 8 of 16 | 31 / 129 | **29 of 80** |
| p2 | 4 of 16 | 6 / 154 | 6 of 80 |

The engine's own hit line moves the same way (p1 `lru/miss`: CPU 73/16 → VK8 54/14 →
VK4 65/15 → VK16 0/13), which is a coarser, slot-level counter and only confirms the tier
took calls away from the CPU cache.

### What this does and does not show
- It fills the cell where matrix-core absence and the Vulkan kernel are no longer
  confounded: **no matrix cores, Vulkan expert kernel, mixed placement, exact.** With
  gfx1151 also exact over Vulkan, backend is the variable left standing, which is
  consistent with the divergence sitting in the HIP path taken by cards without
  matrix cores (#681). It says nothing about HIP directly; I have no HIP host.
- Only the routed-expert kernel is tested. I also ran `COLI_VK_DENSE=1 COLI_VK_ATTN=1`
  arms, but this fixture's dense/shared/attention weights are F32 and `VK_FMT_OK`
  admits only quantized formats, so nothing was preloaded and those arms duplicate the
  expert-only ones. Not evidence about the dense or attention paths.
- Tiny model, two prompts, ~50 positions each. tok/s is meaningless at this size.

Your cheap discriminator still stands as the next step on gfx1030: the 3-way matrix with
`.coli_usage` restored before each run. If GPU-vs-GPU becomes identical there while
GPU-vs-CPU still differs, it is a per-kernel accumulation difference in the HIP fallback,
and this datapoint says Polaris-class silicon is not the reason.

**Comment by huppiflupp** [comment]
The fourth cell, **matrix cores + HIP**: gfx1151 (Strix Halo, RDNA3.5), colibri `dev` @ `5a0b725`, `make colibri HIP=1 HIP_ARCH=gfx1151` (rocWMMA present, so `COLI_GPU_HAS_WMMA=1`). Same fixture and protocol as @crichalchemist: `make_glm_oracle.py --fmt4`, `.coli_usage` restored before every run (16 rows, distinct counts, total 2296), `COLI_TEMP=0 DRAFT=0 USAGE_SAVE=0`. One difference: on the CUDA/HIP path the tier is filled by `pin_load()`, which only runs with AUTOPIN (≥ 5000 selections) or an explicit `PIN=`, so I passed `PIN=<the same file>` and sized the tier with `CUDA_EXPERT_GB` (16 / 8 / 4 experts).

| prompt | arms | md5 of token line | match |
|---|---|---|---|
| p1 (stock, 20) | CPU ×2, GPU16, GPU8 ×2, GPU4 ×2 | `4f65aafa299f…` | 20/20 vs ref, all arms |
| p2 (14-token prompt, 29 new) | same 7 arms | `d632ea34f29b…` | identical across arms |

Mixed placement confirmed: `[PIN] placement: 8 VRAM + 8 RAM`, 43 of 89 expert calls served from VRAM (p1, GPU8).

**One thing this changes about the reading:** under HIP the routed experts never take a WMMA kernel by default. `grouped_s4_wmma` needs `COLI_CUDA_TC_INT4=1` *and* `COLI_GPU_HAS_S4_WMMA`, which is 0 under HIP (rocWMMA has no s4), so gfx1151 runs the **same fallback expert kernel** as gfx1030. So matrix-core presence isn't the variable for the expert path, and neither is the HIP backend as such: the same HIP source is exact on RDNA3.5 and diverges on RDNA2. What's left is gfx1030 code generation for that kernel (dot-product instructions, reduction order, FMA contraction). A diff of `--save-temps` ISA for gfx1030 vs gfx1151, or comparing that kernel's output against CPU on the gfx1030, would settle it.
